File: code_samples/section4/lesson/section4.cpp
#include <vector>
#include <stdexcept>
#include <queue>
#include <stack>
#include <string>
#include <iostream>

// Defines a stack of integers
class IntStack {
public:
    // Add value to the stack
    void push(int value) {
        data_.push_back(value);
    }

    // Remove value from the stack
    int pop() {
        if (data_.empty()) {
            throw std::runtime_error("stack underflow");
        }
        int value = data_.back();
        data_.pop_back();
        return value;
    }

    // Check if the stack is empty
    bool empty() const {
        return data_.empty();
    }

private:
    // The data stored in the stack
    std::vector<int> data_;
};

// Defines a queue of integers
class IntQueue {
public:
    // Add value to the queue
    void enqueue(int value) {
        q_.push(value);
    }

    // Remove value from the queue
    int dequeue() {
        if (q_.empty()) {
            throw std::runtime_error("queue underflow");
        }
        int value = q_.front();
        q_.pop();
        return value;
    }

    // Check if the queue is empty
    bool empty() const {
        return q_.empty();
    }

private:
    // The data stored in the queue
    std::queue<int> q_;
};

// Check if two parentheses are matching
bool isMatching(char open, char close) {
    return (open == '(' && close == ')') ||
           (open == '[' && close == ']') ||
           (open == '{' && close == '}');
}

// Check if a string of parentheses is valid
bool isValidParentheses(const std::string& s) {
    std::stack<char> st;
    for (char c : s) {
        if (c == '(' || c == '[' || c == '{') {
            st.push(c);
        } else if (c == ')' || c == ']' || c == '}') {
            if (st.empty()) return false;
            char open = st.top();
            st.pop();
            if (!isMatching(open, c)) return false;
        }
    }
    return st.empty();
}

/**
 * This function implements the Breadth-first search algorithm.
 * It starts from a given node and explores all the connected nodes in a graph.
 * The graph is represented by an adjacency list.
 *
 * @param start: The starting node for the BFS traversal.
 * @param graph: The adjacency list representing the graph.
 */
void bfs(int start, const std::vector<std::vector<int>>& graph) {
    int n = (int)graph.size();
    std::vector<bool> visited(n, false);
    std::queue<int> q;

    // Mark the start node as visited and enqueue it
    visited[start] = true;
    q.push(start);

    // Print the BFS traversal starting from the start node
    std::cout << "BFS starting from " << start << ":\n";

    while (!q.empty()) {
        // Dequeue a node from the queue and print it
        int node = q.front();
        q.pop();
        std::cout << "  visiting " << node << "\n";

        // Look at all possible neighbors of 'node'
        for (int neighbor : graph[node]) {
            // If there's an edge from node to neighbor and neighbor is not visited
            if (!visited[neighbor]) {
                // Mark the neighbor as visited and enqueue it
                visited[neighbor] = true;
                q.push(neighbor);
            }
        }
    }
}

int main() {
    std::cout << "==== TESTING IntStack ====\n";
    IntStack stack;

    std::cout << "Pushing 10, 20, 30...\n";
    stack.push(10);
    stack.push(20);
    stack.push(30);

    std::cout << "Popping values:\n";
    while (!stack.empty()) {
        int v = stack.pop();
        std::cout << "  popped: " << v << "\n";
    }

    std::cout << "\n==== TESTING IntQueue ====\n";
    IntQueue q;

    std::cout << "Enqueuing 1, 2, 3...\n";
    q.enqueue(1);
    q.enqueue(2);
    q.enqueue(3);

    std::cout << "Dequeuing values:\n";
    while (!q.empty()) {
        int v = q.dequeue();
        std::cout << "  dequeued: " << v << "\n";
    }

    std::cout << "\n==== TESTING isValidParentheses ====\n";

    std::vector<std::string> tests = {
        "()",
        "([])",
        "{[()]}",
        "([)]",
        "((())",
        "abc(def[ghi]{jkl})",
        "",
        "{[}]"
    };

    for (const auto& s : tests) {
        std::cout << "Test \"" << s << "\": "
                  << (isValidParentheses(s) ? "valid" : "invalid")
                  << "\n";
    }

    std::cout << "\n==== TESTING bfs ====\n";

    // Graph:
    // 0 -- 1
    // |  /
    // 2
    std::vector<std::vector<int>> graph(3);
    graph[0] = {1, 2};
    graph[1] = {0, 2};
    graph[2] = {0, 1};

    std::cout << "BFS starting from node 0:\n";
    bfs(0, graph);

    std::cout << "\n==== ALL TESTS COMPLETE ====\n";

    return 0;
}

File: code_samples/section4/lesson/section4.cs
// Check if two parentheses are matching
static bool IsMatching(char open, char close) {
    return (open == '(' && close == ')') ||
           (open == '[' && close == ']') ||
           (open == '{' && close == '}');
}

// Check if a string of parentheses is valid
static bool IsValidParentheses(string s) {
    var stack = new System.Collections.Generic.Stack<char>();
    foreach (var c in s) {
        if (c == '(' || c == '[' || c == '{') {
            stack.Push(c);
        } else if (c == ')' || c == ']' || c == '}') {
            if (stack.Count == 0) return false;
            var open = stack.Pop();
            if (!IsMatching(open, c)) return false;
        }
    }
    return stack.Count == 0;
}

/**
 * This function implements the Breadth-first search algorithm.
 * It starts from a given node and explores all the connected nodes in a graph.
 * The graph is represented by an adjacency list.
 *
 * @param start: The starting node for the BFS traversal.
 * @param graph: The adjacency list representing the graph.
 */
static void Bfs(int start, System.Collections.Generic.List<System.Collections.Generic.List<int>> graph) {
    int n = graph.Count;
    var visited = new bool[n];
    var queue = new System.Collections.Generic.Queue<int>();

    // Mark the start node as visited and enqueue it
    visited[start] = true;
    queue.Enqueue(start);

    // Print the BFS traversal starting from the start node
    Console.WriteLine("BFS starting from " + start + ":");

    while (queue.Count > 0) {
        // Dequeue a node from the queue and print it
        int node = queue.Dequeue();
        Console.WriteLine("  visiting " + node);

        // Look at all possible neighbors of 'node'
        foreach (var neighbor in graph[node]) {
            // If there's an edge from node to neighbor and neighbor is not visited
            if (!visited[neighbor]) {
                // Mark the neighbor as visited and enqueue it
                visited[neighbor] = true;
                queue.Enqueue(neighbor);
            }
        }
    }
}

// ==== TESTING IntStack ====
System.Console.WriteLine("==== TESTING IntStack ====");
var stack = new IntStack();

System.Console.WriteLine("Pushing 10, 20, 30...");
stack.Push(10);
stack.Push(20);
stack.Push(30);

System.Console.WriteLine("Popping values:");
while (!stack.IsEmpty()) {
    int v = stack.Pop();
    System.Console.WriteLine($"  popped: {v}");
}

// ==== TESTING IntQueue ====
System.Console.WriteLine();
System.Console.WriteLine("==== TESTING IntQueue ====");
var q = new IntQueue();

System.Console.WriteLine("Enqueuing 1, 2, 3...");
q.Enqueue(1);
q.Enqueue(2);
q.Enqueue(3);

System.Console.WriteLine("Dequeuing values:");
while (!q.IsEmpty()) {
    int v = q.Dequeue();
    System.Console.WriteLine($"  dequeued: {v}");
}

// ==== TESTING IsValidParentheses ====
System.Console.WriteLine();
System.Console.WriteLine("==== TESTING IsValidParentheses ====");

var tests = new[] {
    "()",
    "([])",
    "{[()]}",
    "([)]",
    "((())",
    "abc(def[ghi]{jkl})",
    "",
    "{[}]"
};

foreach (var s in tests) {
    System.Console.WriteLine(
        $"Test \"{s}\": {(IsValidParentheses(s) ? "valid" : "invalid")}"
    );
}

// ==== TESTING Bfs ====
System.Console.WriteLine();
System.Console.WriteLine("==== TESTING Bfs ====");

// Graph:
// 0 -- 1
// |  /
// 2
var graph = new System.Collections.Generic.List<System.Collections.Generic.List<int>>();
graph.Add(new System.Collections.Generic.List<int> { 1, 2 }); // neighbors of 0
graph.Add(new System.Collections.Generic.List<int> { 0, 2 }); // neighbors of 1
graph.Add(new System.Collections.Generic.List<int> { 0, 1 }); // neighbors of 2

System.Console.WriteLine("BFS starting from node 0:");
Bfs(0, graph);

System.Console.WriteLine();
System.Console.WriteLine("==== ALL TESTS COMPLETE ====");

#pragma warning disable CA1050 // Declare types in namespaces
// Defines a stack of integers
public class IntStack {
    // The stack is implemented as a list
    private readonly System.Collections.Generic.List<int> _data = [];

    // Add value to the stack
    public void Push(int value) {
        _data.Add(value);
    }

    // Remove value from the stack
    public int Pop() {
        if (_data.Count == 0) {
            throw new System.InvalidOperationException("stack underflow");
        }
        int lastIndex = _data.Count - 1;
        int value = _data[lastIndex];
        _data.RemoveAt(lastIndex);
        return value;
    }

    // Check if the stack is empty
    public bool IsEmpty() {
        return _data.Count == 0;
    }
}

// Defines a queue of integers
public class IntQueue {
    // The queue is implemented as a queue
    private readonly System.Collections.Generic.Queue<int> _queue = [];

    // Add value to the queue
    public void Enqueue(int value) {
        _queue.Enqueue(value);
    }

    // Remove value from the queue
    public int Dequeue() {
        if (_queue.Count == 0) {
            throw new System.InvalidOperationException("queue underflow");
        }
        return _queue.Dequeue();
    }

    // Check if the queue is empty
    public bool IsEmpty() {
        return _queue.Count == 0;
    }
}
#pragma warning restore CA1050 // Declare types in namespaces
